docs/proposals/09-network-peering.md

```diff
@@ -0,0 +1,174 @@
+
+* Allow two `Network`s to communicate with each other without exposing their
+  routes / addresses publicly and by only using network routing to do so
+  (i.e. no load-balancing).
+* Be less resource-intensive than comparable solutions like load-balancing /
+  VPN overlays.
+
+### Non-Goals
+
+* Map / transform IP addresses: The peered networks will be interleaved with
+  each other without any transformation. The owners of the networks are
+  responsible for keeping the addresses conflict-free. If there are conflicts,
+  it is up to the networking implementation how to resolve them (e.g. use
+  a 'local-first' approach for which address / route to use).
+
+## Proposal
+
+Extend the `networking.api.onmetal.de.Network` resource with a `spec.peerings`
+field that specifies the desired network peerings and a `status.peerings` that
+reflects the status of these peerings.
+
+A peering has a `name` as handle & primary key (used in `StrategicMergePatch` and
+`Apply`). It references the network to peer with via a `networkRef` field. This
+`networkRef` contains the `name` and the `uid` of the target network. If the `uid`
+is unset, the `Network` controller sets this to the `uid` of the corresponding
+network upon first reconciliation. This ensures that the same object instances
+are peered together by verifying the object identity.
+
+Both `Network`s have to specify a matching peering item (i.e. reference each
+other via `networkRef`) to mutually accept the peering.
+
+The (binding) `phase` of a `spec.peerings` item is reflected in a corresponding
+`status.peerings` item with the same `name`. The `phase` can either be `Pending`,
+meaning there is no active peering, or `Bound`, meaning the peering as described
+in the `spec.peerings` item is in place. The `lastTransitionTime` field is updated
+every time there is a change in the `phase`, allowing users and external
+controllers to determine whether a binding is hanging and to manually delete it
+if necessary.
+
+Example Manifests:
+
+```yaml
+apiVersion: networking.api.onmetal.de/v1alpha1
+kind: Network
+metadata:
+  name: my-network-1
+  uid: 2020dcf9-e030-427e-b0fc-4fec2016e73a
+spec:
+  peerings:
+  # name is the name of the peering configuration
+  - name: my-network-peering
+    networkRef:
+      name: my-network-2
+      # If unset, uid will be filled in by the Network controller
+      # uid: 3030dcf9-f031-801b-f0f0-4fec2016e73a
+status:
+  peerings:
+  - name: my-network-peering
+    # The phase shows the binding progress between two networks.
+    # The initial state is 'Pending' until both peers accept.
+    phase: Bound
+    lastPhaseTransitionTime: "2023-02-16T15:06:58Z"
+---
+apiVersion: networking.api.onmetal.de/v1alpha1
+kind: Network
+metadata:
+  name: my-network-2
+  uid: 3030dcf9-f031-801b-f0f0-4fec2016e73a
+spec:
+  # Both networks have to have the semantically same peering configuration.
+  peerings:
+  - name: my-network-peering
+    networkRef:
+      name: my-network-2
```

Comment:
I think it should be referring to `my-network-1 ` network as per above statement - "(i.e. reference each
other via `networkRef`).

Reply:
Thanks for the pointer, fixed it!